### PR TITLE
Add groupBy

### PR DESCRIPTION
**Pull request by julienrf**
# Summary

- Add a `groupBy` method to `Iterable`
    - Three implementations: a generic implementation in `Buildable` (copied from the current collections), and custom implementations in `View` and `LazyList`
- Add `getOrElse` method to `Map`
- Add `getOrElseUpdate` method to `mutable.Map`
- Add a `newBuilder()` method to appropriate collection factories

# Discussion

The current (I mean, in the current collections) implementation of `groupBy` relies on an available `newBuilder` method.

In our strawman we didn’t have such builders so I wanted to experiment with different implementations.

I compared three implementations:
 - one based on builders (current implementation)
 - one for immutable collections, which depends on an `empty` and a `cons` method to build the groups
 - one for mutable collections, based on `Growable` to build the groups

I compared the implementations with collections (`List`, `ImmutableArray`, `HashSet` and `ArrayBuffer`) of various sizes (from 1 to 7,312,102), containing elements of type `Long`. The benchmark calls the `groupBy` method with the following function `x => x % 5`. (So, the result is a `Map` of 5 groups of equivalent size)

The benchmarks show that for immutable collections with more than 7 elements it’s faster to go with the current implementation.

For mutable collections, the approach based on `Growable`s has the same performance as the current implementation, and it has even slightly better performance on small collections (less than 4 elements).

Therefore, I decided to keep the same implementation as in the current collections (based on builders). We could override the implementation of mutable collections if you think it is worth it…

For reference, here are the different `groupBy` implementations, followed by the code of the benchmarks and the charts.

~~~ scala
// Implementation similar to what we have in the current collections
def groupByBuilder[A, K, C <: Iterable[A]](
  as: C
)(
  newBuilder: () => mutable.Builder[A, C]
)(
  f: A => K
): immutable.Map[K, C] = {
  val m = mutable.Map.empty[K, mutable.Builder[A, C]]
  for (elem <- as) {
    val key = f(elem)
    val bldr = m.getOrElseUpdate(key, newBuilder())
    bldr += elem
  }
  var result = immutable.Map.empty[K, C]
  m.foreach { case (k, v) =>
    result = result + ((k, v.result))
  }
  result
}

// Alternative implementation for immutable collections
// Note that the `cons` function can either prepend or
// append the new value (we don’t guarantee to preserve
// the order of elements within groups)
def groupByImmutable[A, K, C <: Iterable[A]](
  as: C
)(
  empty: C, cons: (A, C) => C
)(
  f: A => K
): immutable.Map[K, C] = {
  var result = immutable.Map.empty[K, C]
  for (elem <- as) {
    val key = f(elem)
    val values = cons(elem, result.getOrElse(key, empty))
    result = result + ((key, values))
  }
  result
}

// Alternative implementation for growable collections
def groupByGrowable[A, K, C <: Iterable[A] with mutable.Growable[A]](
  as: C
)(
  empty: () => C
)(
  f: A => K
): immutable.Map[K, C] = {
  var result = immutable.Map.empty[K, C]
  for (elem <- as) {
    val key = f(elem)
    result.get(key) match {
      case None =>
        val values = empty()
        values += elem
        result = result + ((key, values))
      case Some(values) =>
        values += elem
    }
  }
  result
}
~~~

~~~ scala
object ListGroupByImmutable {
  @Benchmark
  def listGroupBy(bh: Blackhole): Unit = {
    val result = GroupBys.groupByImmutable[Long, Long, List[Long]](xs)(Nil, _ :: _)(_ % 5)
    bh.consume(result)
  }
}

object ListGroupByBuilder {
  @Benchmark
  def listGroupBy(bh: Blackhole): Unit = {
    val result = GroupBys.groupByBuilder[Long, Long, List[Long]](xs)(() => List.newBuilder())(_ % 5)
    bh.consume(result)
  }
}
~~~

listgroupby

immutablearraygroupby

hashsetgroupby

arraybuffergroupby

mutablehashsetgroupby

### CODE REVIEW

File: src/main/scala/strawman/collection/View.scala [hunk 1]
```diff
@@ -13,6 +15,20 @@ trait View[+A] extends Iterable[A] with IterableOps[A, View, View[A]] {
...
+    for (elem <- coll) {
+      val key = f(elem)
+      val bldr = m.getOrElseUpdate(key, ArrayBuffer.empty)
+      bldr += elem
+    }
+    var result = immutable.Map.empty[K, View[A]]
+    m.foreach { case (k, v) =>
+      result = result + ((k, v.view))
+    }
+    result
```
Comment (julienrf) [inline src/main/scala/strawman/collection/View.scala]: This is the implementation of `groupBy` in `View`: each group is represented as an `ArrayBuffer`, and eventually we get `view`s of them.

File: src/main/scala/strawman/collection/View.scala [hunk 2]
```diff
@@ -13,6 +15,20 @@ trait View[+A] extends Iterable[A] with IterableOps[A, View, View[A]] {
     fromIterable(coll)
 
   override def className = "View"
+
+  def groupBy[K](f: (A) => K): immutable.Map[K, View[A]] = {
+    val m = mutable.Map.empty[K, ArrayBuffer[A]]
+    for (elem <- coll) {
+      val key = f(elem)
+      val bldr = m.getOrElseUpdate(key, ArrayBuffer.empty)
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/View.scala]: I think a normal `ArrayBuffer` allocates too much memory.  If there isn't much redundancy you could end up increasing the memory usage by about 5x (!).  I am not sure we have a great alternative data structure right now, but we should think about one.  At the very least I think we should have an array builder and a view of the built array so at *least* we can reclaim all the extra memory after construction is complete.

File: src/main/scala/strawman/collection/immutable/BitSet.scala [hunk 3]
```diff
@@ -66,6 +67,13 @@ object BitSet extends SpecificIterableFactory[Int, BitSet] {
       case _          => empty.++(it)
     }
 
+  def empty: BitSet = new BitSet1(0L)
+
+  def newBuilder(): Builder[Int, BitSet] =
+    new ImmutableBuilder[Int, BitSet](empty) {
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/immutable/BitSet.scala]: This is a performance bug waiting to happen.  Larger immutabe bitsets should absolutely not be built this way; they should be built with a mutable bitset, and upon asking for the result the array should be handed over and/or copied.

File: src/main/scala/strawman/collection/immutable/HashMap.scala [hunk 4]
```diff
@@ -28,8 +30,9 @@ import java.lang.{Integer, String, System}
 @SerialVersionUID(2L)
 sealed trait HashMap[K, +V]
   extends Map[K, V]
-     with MapOps[K, V, HashMap, HashMap[K, V]]
-     with Serializable {
+    with MapOps[K, V, HashMap, HashMap[K, V]]
+    with Buildable[(K, V), HashMap[K, V]]
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/immutable/HashMap.scala]: Why doesn't `MapOps[K, V, HashMap, HashMap[K, V]]` imply `Buildable[(K, V), HashMap[K, V]]`?

File: src/main/scala/strawman/collection/immutable/HashSet.scala [hunk 5]
```diff
@@ -25,15 +25,18 @@ import java.lang.Integer
 @SerialVersionUID(2L)
 sealed trait HashSet[A]
   extends Set[A]
-     with SetOps[A, HashSet, HashSet[A]]
-     with Serializable {
+    with SetOps[A, HashSet, HashSet[A]]
+    with Buildable[A, HashSet[A]]
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/immutable/HashSet.scala]: Same question about `SetOps` implying `Buildable`

File: src/main/scala/strawman/collection/immutable/ListSet.scala [hunk 6]
```diff
@@ -124,5 +127,10 @@ object ListSet extends IterableFactory[ListSet] {
 
   def empty[A]: ListSet[A] = EmptyListSet.asInstanceOf[ListSet[A]]
 
+  def newBuilder[A](): Builder[A, ListSet[A]] =
+    new ImmutableBuilder[A, ListSet[A]](empty) {
+      def add(elem: A): this.type = { elems = elems + elem; this }
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/immutable/ListSet.scala]: Any way to push this implementation up into the trait without losing performance?  There's a lot of repetition of this pattern.

File: src/main/scala/strawman/collection/mutable/ListBuffer.scala [hunk 7]
```diff
@@ -12,7 +12,7 @@ import scala.Predef.{assert, intWrapper}
 
 /** Concrete collection type: ListBuffer */
 class ListBuffer[A]
-  extends Seq[A]
+  extends GrowableSeq[A]
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/mutable/ListBuffer.scala]: Why do we need to separately say this is growable, buildable, and a builder?  Also, can't every mutable collection be treated as its own builder?

File: src/main/scala/strawman/collection/immutable/BitSet.scala (hunk 3 above)
Comment (julienrf) [inline src/main/scala/strawman/collection/immutable/BitSet.scala]: Oh, indeed, definitely!

File: src/main/scala/strawman/collection/immutable/HashMap.scala (hunk 4 above)
Comment (julienrf) [inline src/main/scala/strawman/collection/immutable/HashMap.scala]: I think that the decision of extending `Buildable` can only be made at the leaf level (for each concrete collection type). For instance, `Seq` does not imply `Buildable`, which is good because we don’t want `LazyList` to implement `Buildable`.

File: src/main/scala/strawman/collection/immutable/ListSet.scala (hunk 6 above)
Comment (julienrf) [inline src/main/scala/strawman/collection/immutable/ListSet.scala]: I could make `ImmutableBuilder[A, C]` take as parameter a `(C, A) => C` function defining how to add an element to a given collection. Then at use site that would look like the following:

~~~ scala
new ImmutableBuilder[A, ListSet[A]](empty, _ + _)
~~~

WDYT?

File: src/main/scala/strawman/collection/mutable/ListBuffer.scala (hunk 7 above)
Comment (julienrf) [inline src/main/scala/strawman/collection/mutable/ListBuffer.scala]: > Why do we need to separately say this is growable, buildable, and a builder?

Indeed we could probably assume that `Growable` imply `Buildable`. I see no example where we would have the former without also having the latter.

I decided to put the method that returns the builder in the companion object because I think that could be useful for other purposes (see #97).

> can't every mutable collection be treated as its own builder?

That’s a good question. Actually, @odersky initially designed mutable collections to be their own builders. I thought that it would make the hierarchy slightly more complicated (that’s yet another type that shows up in the linearization) and also I think the public `result` method of builders makes little sense on a concrete collection.

Comment (julienrf) [inline src/main/scala/strawman/collection/mutable/ListBuffer.scala]: > Why do we need to separately say this is growable, buildable

Actually, in the current design `Growable[A]` takes just one type parameter (the type of the elements). If we want to make `Growable[A]` imply `Buildable[A, C]` then we need to add the `C` type parameter to `Growable` and on all its subclasses. A consequence is that this `C` should be supplied by a `XxxOps` trait, which means that we would have a `GrowableIterableOps` branch in the hierarchy that would be refined for `Set`, `Map` and `Seq`. That would be a lot of additional traits :(

Or, could we assume that all mutable collections are `Growable`? (in such a case we could make `mutable.IterableOps` extend `Buildable` and get rid of the `mutable.Iterable` / `mutable.GrowableIterable` distinction)

Comment (Ichoran) [inline src/main/scala/strawman/collection/mutable/ListBuffer.scala]: I forgot that Arrays are not growable.  I guess there is a distinction there to maintain.

File: src/main/scala/strawman/collection/immutable/ListSet.scala (hunk 6 above)
Comment (Ichoran) [inline src/main/scala/strawman/collection/immutable/ListSet.scala]: That's not much better, and we want to be really careful about per-element overhead.  I was thinking about pushing it up into a `SetFactory` which extends `IterableFactory`.  Not sure if there will be performance consequences, though.

File: src/main/scala/strawman/collection/immutable/HashMap.scala (hunk 4 above)
Comment (Ichoran) [inline src/main/scala/strawman/collection/immutable/HashMap.scala]: So it's just because someone might write `MapOps[K, V, Map, Map[K, V]]` that we don't want to be `Buildable`?

**Comment by julienrf** [comment]
@Ichoran can you summarize what needs to be changed?

**Comment by julienrf** [comment]
I rebased the PR and implemented `groupBy` in `Range` and `NumericRange`. I moved the actual implementation from `Buildable` to a `collection.generic.GroupBy` object so that I could reuse it even in collections types that do not extend `Buildable` (such as `Range` and `NumericRange`)

File: src/main/scala/strawman/collection/immutable/HashMap.scala (hunk 4 above)
Comment (szeiger) [inline src/main/scala/strawman/collection/immutable/HashMap.scala]: For `Map` it probably makes sense to extend `Buildable`, just like for `Set` (which I am working on at the moment)

**Comment by Ichoran** [comment]
@julienrf - I'll try to check it this evening.

**Comment by Ichoran** [comment]
Hrm, not sure if there's a way to "unreview".  Dismissing the review doesn't seem to be what I wanted.

**Comment by julienrf** [comment]
Superseded by #111.
